src/rpc/server.zig

```diff
@@ -0,0 +1,547 @@
+        read_buffer_size: u32,
+        /// The socket address to listen on for incoming HTTP and/or RPC requests.
+        socket_addr: std.net.Address,
+
+        pub const MIN_READ_BUFFER_SIZE = 256;
+    };
+
+    pub const InitError = std.net.Address.ListenError;
+
+    /// The returned result must be pinned to a memory location before calling any methods.
+    pub fn init(params: InitParams) std.net.Address.ListenError!Server {
+        var tcp_server = try params.socket_addr.listen(.{
+            // NOTE: ideally we would be doing this nonblockingly, however this doesn't work properly on mac,
+            // so for testing purposes we can't teste the `serve` functionality directly.
+            .force_nonblocking = false,
+        });
+        errdefer tcp_server.deinit();
+
+        return .{
+            .allocator = params.allocator,
+            .logger = params.logger.withScope(logger_scope),
+
+            .snapshot_dir = params.snapshot_dir,
+            .latest_snapshot_gen_info = params.latest_snapshot_gen_info,
+
+            .wait_group = .{},
+            .thread_pool = params.thread_pool,
+
+            .read_buffer_size = @max(params.read_buffer_size, InitParams.MIN_READ_BUFFER_SIZE),
+            .tcp = tcp_server,
+        };
+    }
+
+    /// Blocks until all tasks are completed, and then closes the server.
+    /// Does not force the server to exit.
+    pub fn joinDeinit(server: *Server) void {
+        server.wait_group.wait();
+        server.tcp.deinit();
+    }
+
+    /// Spawn the serve loop as a separate thread.
+    pub fn serveSpawn(
+        server: *Server,
+        exit: *std.atomic.Value(bool),
+    ) std.Thread.SpawnError!std.Thread {
+        return std.Thread.spawn(.{}, serve, .{ server, exit });
+    }
+
+    /// Calls `acceptAndServeConnection` in a loop until `exit.load(.acquire)`.
+    pub fn serve(server: *Server, exit: *std.atomic.Value(bool)) !void {
+        while (!exit.load(.acquire)) {
+            _ = try server.acceptAndServeConnection();
+        }
+    }
+
+    pub const AcceptAndServeConnectionError =
+        std.mem.Allocator.Error ||
+        std.http.Server.ReceiveHeadError ||
+        AcceptConnectionError;
+
+    pub fn acceptAndServeConnection(server: *Server) AcceptAndServeConnectionError!void {
+        const conn = (try acceptConnection(&server.tcp, server.logger)).?;
+        errdefer conn.stream.close();
+
+        server.wait_group.start();
+        errdefer server.wait_group.finish();
+
+        const hct_buffer = try server.allocator.alignedAlloc(
+            u8,
+            @alignOf(HandleConnectionTask),
+            HandleConnectionTask.initBufferSize(server.read_buffer_size),
+        );
+        errdefer server.allocator.free(hct_buffer);
+
+        const new_hct = try HandleConnectionTask.initBufferAndReceiveHead(
+            hct_buffer,
+            server,
+            conn,
+        );
+        errdefer new_hct.http_server.connection.stream.close();
```

Comment:
would probs be simpler to have this logic in one `init` method -- its a bit confusing since we dont typically setup thread pool tasks like this